wagtail/core/telepath.py

```diff
@@ -4,7 +4,73 @@
 from wagtail.admin.staticfiles import versioned_static
 
 
-adapters = {}
+DICT_RESERVED_KEYS = ['_type', '_args', '_dict']
+
+
+class BaseAdapter:
+    """Handles serialisation of a specific object type"""
+    def pack(self, obj, context):
+        """
+        Translates obj into serialisable form. Any media declarations that will be required for
+        deserialisation of the object should be passed to context.add_media().
+
+        This base implementation handles simple JSON-serialisable values such as strings, and
+        returns them unchanged.
+        """
+        return obj
+
+
+class DictAdapter(BaseAdapter):
+    """Handles serialisation of dicts"""
+    def pack(self, obj, context):
+        packed_obj = {
+            str(key): context.pack(val)
+            for key, val in obj.items()
+        }
+        if any(reserved_key in packed_obj for reserved_key in DICT_RESERVED_KEYS):
+            # this dict contains keys such as _type that would collide with our object notation,
+            # so wrap it in an explicit _dict to disambiguate
+            return {'_dict': packed_obj}
+        else:
+            return packed_obj
+
+
+class Adapter(BaseAdapter, metaclass=MediaDefiningClass):
+    """
+    Handles serialisation of custom types.
+    Subclasses should define:
+    - js_constructor: namespaced identifier for the JS constructor function that will unpack this
+        object
+    - js_args(obj): returns a list of (telepath-packable) arguments to be passed to the constructor
+    - get_media(obj) or class Media: media definitions necessary for unpacking
+
+    The adapter should then be registered with register(adapter, cls).
+    """
+
+    def get_media(self, obj):
+        return self.media
+
+    def pack(self, obj, context):
+        context.add_media(self.get_media(obj))
+        return {
+            '_type': self.js_constructor,
+            '_args': [context.pack(arg) for arg in self.js_args(obj)]
+        }
+
+
+adapters = {
+    # Primitive value types that are unchanged on serialisation
```

Comment:
Where would be the best place to register adapters for standard library types like dates and Decimals?

Reply:
Same as we've done for Django widgets (wagtail.core.widget_adapters) - some module that gets loaded on startup via the appconf. It would probably make sense for dates (and possibly decimals? A bit rusty on how they work in Python vs JS) to be built in to the base telepath module as standard though.